tsl/src/chunk.c

```diff
@@ -1180,3 +1186,538 @@ chunk_merge_chunks(PG_FUNCTION_ARGS)
+									   splitrel->rd_rel->reltablespace,
+									   splitrel->rd_rel->relam,
+									   relpersistence,
+									   AccessExclusiveLock);
+
+	Relation splitrel_new = table_open(new_heap_relid, AccessExclusiveLock);
+	Relation new_chunkrel = table_open(new_chunk->table_id, AccessExclusiveLock);
+
+	estate = CreateExecutorState();
+
+	/* Create the tuple slot */
+	srcslot = MakeSingleTupleTableSlot(RelationGetDescr(splitrel), table_slot_callbacks(splitrel));
+
+	/* Setup the state we need for each new chunk partition */
+	splitinfos[0] = relation_split_info_create(splitrel, splitrel_new, &cutoffs);
+	splitinfos[1] = relation_split_info_create(splitrel, new_chunkrel, &cutoffs);
+
+	/* Scan through the rows using SnapshotAny to see everything. */
+	scan = table_beginscan(splitrel, SnapshotAny, 0, NULL);
+
+	/*
+	 * Switch to per-tuple memory context and reset it for each tuple
+	 * produced, so we don't leak memory.
+	 */
+	econtext = GetPerTupleExprContext(estate);
+	oldcxt = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
+
+	/*
+	 * Read all the data from the split relation and route the tuples to the
+	 * new partitions. Do some vacuuming and cleanup at the same
+	 * time. Transfer all visibility information to the new relations.
+	 *
+	 * Main loop inspired by heapam_relation_copy_for_cluster() used to run
+	 * CLUSTER and VACUUM FULL on a table.
+	 */
+	double num_tuples = 0.0;
+	double tups_vacuumed = 0.0;
+	double tups_recently_dead = 0.0;
+
+	BufferHeapTupleTableSlot *hslot = (BufferHeapTupleTableSlot *) srcslot;
+
+	while (table_scan_getnextslot(scan, ForwardScanDirection, srcslot))
+	{
+		RelationSplitInfo *rsi = NULL;
+		bool isnull = true;
+		HeapTuple tuple;
+		Buffer buf;
+		bool isdead;
+
+		CHECK_FOR_INTERRUPTS();
+		ResetExprContext(econtext);
+
+		tuple = ExecFetchSlotHeapTuple(srcslot, false, NULL);
+		buf = hslot->buffer;
+
+		LockBuffer(buf, BUFFER_LOCK_SHARE);
+
+		switch (HeapTupleSatisfiesVacuum(tuple, cutoffs.OldestXmin, buf))
+		{
+			case HEAPTUPLE_DEAD:
+				/* Definitely dead */
+				isdead = true;
+				break;
+			case HEAPTUPLE_RECENTLY_DEAD:
+				tups_recently_dead += 1;
+				/* fall through */
+				TS_FALLTHROUGH;
+			case HEAPTUPLE_LIVE:
+				/* Live or recently dead, must copy it */
+				isdead = false;
+				break;
+			case HEAPTUPLE_INSERT_IN_PROGRESS:
+
+				/*
+				 * Since we hold exclusive lock on the relation, normally the
+				 * only way to see this is if it was inserted earlier in our
+				 * own transaction.  However, it can happen in system
+				 * catalogs, since we tend to release write lock before commit
+				 * there.  Give a warning if neither case applies; but in any
+				 * case we had better copy it.
+				 */
+				if (!TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetXmin(tuple->t_data)))
+					elog(WARNING,
+						 "concurrent insert in progress within table \"%s\"",
+						 RelationGetRelationName(splitrel));
+				/* treat as live */
+				isdead = false;
+				break;
```

Comment:
You might want to add an isolation test for this, although I do not expect any problems.